platform/src/server/http/http_server.ts

```diff
@@ -55,41 +86,89 @@ export class HttpServer {
-    // ```
-    const app = this.app as any;
-
-    if (!config.ssl.enabled) {
-      return http.createServer(app);
-    }
+    const options: ServerOptions = {
+      host: config.host,
+      port: config.port,
+      routes: {
+        cors: config.cors,
+        payload: {
+          maxBytes: config.maxPayload.getValueInBytes(),
+        },
+        validate: {
+          options: {
+            abortEarly: false
+          }
+        }
+      },
+      state: {
+        strictHeader: false
+      },
+    };
+
+    const ssl = config.ssl;
+    if (ssl.enabled) {
+      const tlsOptions: TLSOptions = {
+        ca: config.ssl.certificateAuthorities &&
+          config.ssl.certificateAuthorities.map(caFilePath => readFileSync(caFilePath)),
+
+        key: readFileSync(ssl.key!),
+        cert: readFileSync(ssl.certificate!),
+        passphrase: ssl.keyPassphrase,
 
-    // TODO: add support for `secureOptions`.
-    return https.createServer(
-      {
-        key: readFileSync(config.ssl.key!),
-        cert: readFileSync(config.ssl.certificate!),
-        ca:
-          config.ssl.certificateAuthorities &&
-          config.ssl.certificateAuthorities.map(caFilePath =>
-            readFileSync(caFilePath)
-          ),
-        passphrase: config.ssl.keyPassphrase,
         ciphers: config.ssl.cipherSuites.join(':'),
         // We use the server's cipher order rather than the client's to prevent the BEAST attack.
         honorCipherOrder: true,
-      },
-      app
-    );
+        secureOptions: ssl.getSecureOptions(),
+      };
+
+      // TODO: Hapi types have a typo in `tls` property type definition: `https.RequestOptions` is used instead of
+      // `https.ServerOptions`, and `honorCipherOrder` isn't presented in `https.RequestOptions`.
+      options.tls = tlsOptions as any;
+    }
+
+    return new Server(options);
+  }
+
+  private setupBasePathRewrite(server: Server, config: HttpConfig) {
+    if (config.basePath === undefined || !config.rewriteBasePath) {
+      return;
+    }
+
+    const basePath = config.basePath;
+    server.ext('onRequest', (request: Request, responseToolkit: ResponseToolkit) => {
+      const newURL = modifyUrl(request.url.href!, (urlParts) => {
+        if (urlParts.pathname != null && urlParts.pathname.startsWith(basePath)) {
+          urlParts.pathname = urlParts.pathname.replace(basePath, '') || '/';
+        } else {
+          return {};
+        }
+      });
+
+      if (!newURL) {
+        return responseToolkit
+          .response('Not Found')
+          .code(404)
+          .takeover();
+      }
+
+      request.setUrl(newURL);
+      // We should update raw request as well since it can be proxied to the old platform
+      // where base path isn't expected.
+      request.raw.req.url = request.url.href;
+
+      return responseToolkit.continue;
+    });
+  }
+
+  private getRouteFullPath(routerPath: string, routePath: string) {
```

Comment:
__note:__ maybe there is a better way to handle "double-slash" problem, but I didn't find anything nice and simple. Happy to hear any suggestions here.

Reply:
This double-slash issue is introduced because of ExpressJS treating `/foo` and `/foo/` the same, but Hapi treats them differently?

Reply:
Nope, the "trailing-slashes" issue is the different one :). Here I just wanted to say that with ExpressJS Router we didn't have to deal with concatenating of router and route URLs directly, it was handled by ExpressJS Router internally. But to define a Hapi route we should provide the path that already includes router's path/prefix.